core/device_impl.cpp

```diff
@@ -461,13 +484,36 @@ MavlinkCommands::Result DeviceImpl::set_flight_mode(uint8_t custom_sub_mode, uin
     return ret;
 }
 
-void DeviceImpl::set_flight_mode_async(uint8_t custom_sub_mode, uint8_t custom_mode,
+void DeviceImpl::set_flight_mode_async(FLIGHT_MODE device_mode,
                                        command_result_callback_t callback)
 {
-    uint8_t flag_safety_armed = DeviceImpl::is_armed() ? MAV_MODE_FLAG_SAFETY_ARMED : 0;
+    uint8_t flag_safety_armed = is_armed() ? MAV_MODE_FLAG_SAFETY_ARMED : 0;
 
     uint8_t mode = MAV_MODE_FLAG_CUSTOM_MODE_ENABLED | flag_safety_armed;
 
+    uint8_t custom_mode = px4::PX4_CUSTOM_MAIN_MODE_AUTO;
+    uint8_t custom_sub_mode = 0;
+
+    switch (device_mode) {
+        case HOLD: custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_LOITER;
+            break;
+        case RETURN_TO_LAUNCH :
+            custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_RTL;
+            break;
+        case TAKEOFF : custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_TAKEOFF;
+            break;
+        case LAND : custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_LAND;
+            break;
+        case MISSION : custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_MISSION;
+            break;
+        case FOLLOW_ME : custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_FOLLOW_TARGET;
+            break;
+        case OFFBOARD : custom_mode = px4::PX4_CUSTOM_MAIN_MODE_OFFBOARD;
+            break;
+        default : return ;
```

Comment:
We need to handle for unsupported mode, like it is handled [here](https://github.com/Rjasuja/DroneCore/blob/5e9b9703ece19bcce112a5be068c57442651baa8/core/device_impl.cpp#L582).